functions/types.go

```diff
@@ -0,0 +1,124 @@
+package functions
+
+import (
+	substraitgo "github.com/substrait-io/substrait-go"
+	"github.com/substrait-io/substrait-go/types"
+	"strings"
+)
+
+var nameToTypeMap = map[string]types.Type{
+	"boolean":       &types.BooleanType{},
+	"i8":            &types.Int8Type{},
+	"i16":           &types.Int16Type{},
+	"i32":           &types.Int32Type{},
+	"i64":           &types.Int64Type{},
+	"fp32":          &types.Float32Type{},
+	"fp64":          &types.Float64Type{},
+	"binary":        &types.BinaryType{},
+	"string":        &types.StringType{},
+	"timestamp":     &types.TimestampType{},
+	"date":          &types.DateType{},
+	"time":          &types.TimeType{},
+	"timestamp_tz":  &types.TimestampTzType{},
+	"interval_year": &types.IntervalYearType{},
+	"interval_day":  &types.IntervalDayType{},
+	"uuid":          &types.UUIDType{},
+
+	// short names
+	"bool":  &types.BooleanType{},
+	"vbin":  &types.BinaryType{},
+	"str":   &types.StringType{},
+	"ts":    &types.TimestampType{},
+	"tstz":  &types.TimestampTzType{},
+	"iyear": &types.IntervalYearType{},
+	"iday":  &types.IntervalDayType{},
+}
+
+func isSupportedType(typeString string) bool {
+	_, ok := nameToTypeMap[typeString]
+	return ok
+}
+
+type typeRegistryImpl struct {
+	typeMap map[string]types.Type
+}
+
+func NewTypeRegistry() TypeRegistry {
+	return &typeRegistryImpl{typeMap: nameToTypeMap}
+}
+
+func (t *typeRegistryImpl) GetTypeFromTypeString(typeString string) (types.Type, error) {
+	return getTypeFromTypeString(typeString, t.typeMap)
+}
+
+func getTypeFromTypeString(typeString string, typeMap map[string]types.Type) (types.Type, error) {
+	// TODO handle parameterized types
+	nullable := types.NullabilityRequired
+	if strings.HasSuffix(typeString, "?") {
+		typeString = typeString[:len(typeString)-1]
+		nullable = types.NullabilityNullable
+	}
+	if typ, ok := typeMap[typeString]; ok {
+		return typ.WithNullability(nullable), nil
+	}
+	return nil, substraitgo.ErrNotFound
+}
+
+type localTypeRegistryImpl struct {
+	nameToType      map[string]types.Type
+	localNameToType map[string]types.Type
+	typeInfoMap     map[string]typeInfo
+}
+
+type typeInfo struct {
+	typ               types.Type
+	shortName         string
+	localName         string
+	supportedAsColumn bool
+}
+
+func NewLocalTypeRegistry(typeInfos []typeInfo) LocalTypeRegistry {
```

Comment:
We can have another two maps to convert from "name" dialect to substrait and vice versa basically
map[string(localname)] --> string (substraitName)
map[string(substraitName)] --> string (localname)
and existing map to give typeInfo and type. 

so finally we will have below maps
1. map[string(substraitName)] --> types.Type (This is in memory Substrait type so substrait map represents it)
2. map[string(substraitName)]  --> typeInfo (Substrait type info)
3. map[string(localname)]  --> typeInfo (local type info)
4. map[string(localname)] --> string (substraitName)
5. map[string(substraitName)] --> string (localname)

Reply:
2 is not needed. Result of [this](https://github.com/substrait-io/substrait-go/pull/32#issuecomment-2275364351) comment. 

It was because I felt it is easier this way to read. But name of map captures its key type (local or substrait). So a matter of choice. Feel free to ignore this comment